ASoC: SOF: Intel: reserve link DMA for sdw bpt stream

bardliao opened the pull request:
Currently, hda_sdw_bpt_dma_prepare() get a HDA stream and use the link DMA but doesn't reserve it. It works fine because
we assume the SwoundWire BPT will not run with audio streams simultaneously. Create and use the new helpers to reserve the link DMA and allow running BTP and audio stream simultaneously.

bardliao committed:
ASoC: SOF: Intel: add hda_dsp_stream_pair_get/put helpers

Currently, hda_dsp_stream_get/put are used to get/put the host dma.
However, we may want to use a hda stream that both host and link dma are
available. Add helper to find the hda stream and reserve/release it.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1211,3 +1211,114 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+		dev_err(sdev->dev, "error: no stream available\n");
+		return ERR_PTR(-ENODEV);
+	}
+	hstream = &hext_stream->hstream;
```
`s` _is_ `hstream` , no? Probably use only `hstream` and remove `s`?

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1211,3 +1211,114 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+		goto out_put;
+	}
+
+	hstream->period_bytes = 0;/* initialize period_bytes */
```
space before the comment

ranj063 replied:
not resolved yet

bardliao replied:
fixed

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1211,3 +1211,114 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+
+	ret = hda_dsp_stream_hw_params(sdev, hext_stream, dmab, NULL);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: hdac prepare failed: %d\n", ret);
```
no `error:` in prints, but __func__ helps.

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1211,3 +1211,114 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+				 struct hdac_ext_stream *hext_stream)
+{
+	struct snd_sof_dev *sdev =  dev_get_drvdata(dev);
+	struct hdac_stream *hstream = &hext_stream->hstream;
```
we have this helper: `hdac_stream(hext_stream)`, might be better to use it here as  well?

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1211,3 +1211,114 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+
+	hda_dsp_stream_put(sdev, hstream->direction, hstream->stream_tag);
+	snd_hdac_ext_stream_release(hext_stream, HDAC_EXT_STREAM_TYPE_LINK);
+	hstream->running = 0;
```
This is handled by hda_cl_trigger() as the SNDRV_PCM_TRIGGER_STOP is passed to hda_dsp_stream_trigger().
But please check and remove

bardliao replied:
Yes, it is handled by hda_cl_trigger(), but I prefer to keep it as it is not guarantee that hda_cl_trigger() will be called before hda_dma_cleanup().

ujfalusi reviewed (commented):
@bardliao, I think this is much cleaner than the previous version, few comments to address.

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1211,3 +1211,112 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+	hstream->bufsize = 0;
+	hstream->format_val = 0;
+out_put:
+	hda_dsp_stream_put(sdev, direction, hstream->stream_tag);
```
you likely need to do snd_hdac_ext_stream_release() as well?

bardliao replied:
> you likely need to do snd_hdac_ext_stream_release() as well?

Yes, thanks for the reminding.

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1211,3 +1211,113 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
 }
 EXPORT_SYMBOL_NS(hda_dsp_get_stream_ldp, "SND_SOC_SOF_INTEL_HDA_COMMON");
+
+struct hdac_ext_stream*
```
I think a space before the `*` is due. To align with hda_dsp_stream_get()

bardliao replied:
> I think a space before the `*` is due. To align with hda_dsp_stream_get()

Updated.

ranj063 replied:
missing space still?

bardliao replied:
I know what happened. I fixed it previously. Then I removed in the last version and copy the function from hda-loader.c again in this version. Fixed it again now.

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1211,3 +1211,113 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+
+struct hdac_ext_stream *
+hda_dsp_dma_channel_pair_get(struct device *dev, unsigned int format, unsigned int size,
+			     struct snd_dma_buffer *dmab, int direction)
```
Looking at this again, I think the format/size/dmab should not be required to get the channel pair. Strictly speaking from API point of view.

This is how the BPT is using it, yes, but that does not mean that if we have new user (we might never) it will also need this to be done.

Should we restrict this to get the channel pair and add a wrapper in bpt code to request the pair and do the configuration?

@kv2019i, @ranj063, what do you think?

bardliao replied:
It sounds better. Let's make the helpers more generic.

bardliao committed:
ASoC: SOF: Intel: hda-sdw-bpt: support simultaneous audio and BPT streams

Currently the SoundWire BPT stream uses the paired link DMA but not
reserve it. It works without any issue because we assume the SoundWire
BPT will not run with audio streams simultaneously.
To support simultaneous audio and BPT streams, we need to use the
hda_dma_prepare/cleanup helpers to reserve the pair link host DMA.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -264,14 +267,33 @@ hda_dsp_stream_get(struct snd_sof_dev *sdev, int direction, u32 flags)
+	struct hdac_ext_stream *hext_stream = _hda_dsp_stream_get(sdev, direction, flags, true);
+
+	if (hext_stream)
+		hext_stream->link_locked = true;
```
In theory the link could be snatched by other stream between the call _hda_dsp_stream_get() and this point.

bardliao replied:
> In theory the link could be snatched by other stream between the call _hda_dsp_stream_get() and this point.

Fixed, thanks.

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -90,6 +90,89 @@ static int chain_dma_trigger(struct snd_sof_dev *sdev, unsigned int stream_tag,
 	return sof_ipc_tx_message_no_reply(sdev->ipc, &msg, 0);
 }
 
+static struct hdac_ext_stream *
```
Can you fix the typo with BPT in the commit message?

bardliao replied:
fixed

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -90,6 +90,89 @@ static int chain_dma_trigger(struct snd_sof_dev *sdev, unsigned int stream_tag,
 }
 
+static struct hdac_ext_stream *
+hda_sdw_prepare(struct device *dev, unsigned int format, unsigned int size,
```
Is the only difference te use of hda_dsp_stream_pair_get() instead of stream_get() compared to hda_cl_prepare()? Can we reduce code duplication by reusing part of the code?

bardliao replied:
> Is the only difference te use of hda_dsp_stream_pair_get() instead of stream_get() compared to hda_cl_prepare()? Can we reduce code duplication by reusing part of the code?

Yes, updated.

bardliao committed:
ASoC: SOF: Intel: add hda_dma_prepare/cleanup helpers

SoundWire BPT stream needs to use link and host DMAs. Thus we need
helpers to prepare and cleanup the link and host DMAs. Currently the
SoundWire BPT stream uses hda_cl_prepare/cleanup helpers. It works fine
because we assume the SwoundWire BPT will not run with audio streams
simultaneously. The new helpers are copied from hda_cl_prepare/cleanup
and add a flag to reserve the paired host and link DMAs. The new helpers
will be used by both code loader and SoundWire BPT.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -275,38 +218,7 @@ EXPORT_SYMBOL_NS(hda_cl_trigger, "SND_SOC_SOF_INTEL_HDA_COMMON");
-	}
-
-	return ret;
+	return hda_dma_cleanup(dev, dmab, persistent_buffer, hext_stream, false);
```
this patch needs to be squahsed with the previous one

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -118,7 +118,8 @@ static int hda_sdw_bpt_dma_prepare(struct device *dev, struct hdac_ext_stream **
 	dev_dbg(dev, "direction %d format_val %#x\n", direction, format);
 
-	bpt_stream = hda_cl_prepare(dev, format, bpt_num_bytes, dmab_bdl, false, direction, false);
+	bpt_stream = hda_dma_prepare(dev, format, bpt_num_bytes, dmab_bdl,
```
Maybe update the commit title and messages to say "support simultaneous audio and BPT streams"?

ranj063 replied:
and fix typo in Soundwire in the commit message

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1246,3 +1246,118 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+		hext_stream = hda_dsp_stream_get(sdev, direction, 0);
+
+	if (!hext_stream) {
+		dev_err(sdev->dev, "error: no stream available\n");
```
drop the `error:` and replace it with __func__

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1246,3 +1246,118 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+	} else {
+		ret = hda_dsp_stream_hw_params(sdev, hext_stream, dmab, NULL);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: hdac prepare failed: %d\n", ret);
```
and same for the other `error:` in prints.

bardliao replied:
done

ujfalusi reviewed (commented):
only few nitpicks..

ujfalusi commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1246,3 +1246,119 @@ u64 hda_dsp_get_stream_ldp(struct snd_sof_dev *sdev,
+struct hdac_ext_stream *
+hda_dma_prepare(struct device *dev, unsigned int format, unsigned int size,
+		struct snd_dma_buffer *dmab, bool persistent_buffer, int direction,
+		bool is_iccmax, bool pair)
```
The generic naming of hda_dma_prepare() / hda_dma_cleanup() I think is confusing.
This is implementing a special use of HD-DMA for data transfer, it does buffer allocation also.

It handles the CL (Code Loader - pair == false)  and a data transfer channel to external device.

Can we find a better name? (sorry)

bardliao replied:
Yeah, I thought about it. But I can't find a better name at that moment. Maybe we can figure one out together. How about `hda_dt_prepare/cleanup` DT: Data Transfer

ujfalusi replied:
To stick with the existing (not too intuitive) naming convention used within hda-stream.c, what about:
hda_data_stream_prepare() / hda_data_stream_cleanup()
?

bardliao replied:
Thanks, will update.

ujfalusi reviewed (approved):
@bardliao, I think this is a great version!
